File: frontend/src/stores/ofo.js
import { defineStore } from 'pinia'
import { ref, computed } from 'vue'
import api from '../services/api'
import { getLotSize } from '@/constants/trading'
import { useStrategyStore } from './strategy'

/**
 * OFO (Options For Options) Store
 *
 * Manages state for finding and displaying the best option strategy combinations.
 */
export const useOFOStore = defineStore('ofo', () => {
  // =========================================================================
  // STATE
  // =========================================================================

  // Selection state
  const underlying = ref('NIFTY')
  const expiry = ref('')
  const expiries = ref([])

  // Strategy selection (multi-select)
  const selectedStrategies = ref([])

  // Configuration
  const strikeRange = ref(10) // ±strikes from ATM (5, 10, 15, 20)
  const lots = ref(1)

  // Results
  const results = ref({}) // strategy_type -> array of top 3 results
  const spotPrice = ref(0)
  const atmStrike = ref(0)
  const lotSize = ref(getLotSize('NIFTY'))
  const calculationTimeMs = ref(0)
  const totalCombinationsEvaluated = ref(0)
  const lastCalculated = ref(null)

  // Auto-refresh
  const autoRefreshEnabled = ref(false)
  const autoRefreshInterval = ref(5) // minutes
  let autoRefreshTimer = null

  // UI state
  const isLoading = ref(false)
  const error = ref(null)

  // Available strategies (loaded from API)
  const availableStrategies = ref([
    { key: 'iron_condor', name: 'Iron Condor', category: 'neutral', legs_count: 4 },
    { key: 'iron_butterfly', name: 'Iron Butterfly', category: 'neutral', legs_count: 4 },
    { key: 'short_straddle', name: 'Short Straddle', category: 'neutral', legs_count: 2 },
    { key: 'short_strangle', name: 'Short Strangle', category: 'neutral', legs_count: 2 },
    { key: 'long_straddle', name: 'Long Straddle', category: 'volatile', legs_count: 2 },
    { key: 'long_strangle', name: 'Long Strangle', category: 'volatile', legs_count: 2 },
    { key: 'bull_call_spread', name: 'Bull Call Spread', category: 'bullish', legs_count: 2 },
    { key: 'bear_put_spread', name: 'Bear Put Spread', category: 'bearish', legs_count: 2 },
    { key: 'butterfly_spread', name: 'Butterfly Spread', category: 'neutral', legs_count: 4 }
  ])

  // =========================================================================
  // GETTERS
  // =========================================================================

  const selectedCount = computed(() => selectedStrategies.value.length)

  const hasResults = computed(() => Object.keys(results.value).length > 0)

  const orderedResults = computed(() => {
    // Return results in the order of selectedStrategies
    const ordered = {}
    for (const strategyKey of selectedStrategies.value) {
      if (results.value[strategyKey]) {
        ordered[strategyKey] = results.value[strategyKey]
      }
    }
    return ordered
  })

  const lastCalculatedFormatted = computed(() => {
    if (!lastCalculated.value) return null
    const date = new Date(lastCalculated.value)
    return date.toLocaleTimeString('en-IN', {
      hour: '2-digit',
      minute: '2-digit',
      second: '2-digit'
    })
  })

  // =========================================================================
  // ACTIONS
  // =========================================================================

  /**
   * Set underlying and fetch expiries
   */
  async function setUnderlying(ul) {
    underlying.value = ul.toUpperCase()
    lotSize.value = getLotSize(underlying.value)
    expiry.value = ''
    results.value = {}
    lastCalculated.value = null
    await fetchExpiries()
  }

  /**
   * Fetch available expiries for selected underlying
   */
  async function fetchExpiries() {
    try {
      const response = await api.get('/api/options/expiries', {
        params: { underlying: underlying.value }
      })
      expiries.value = response.data.expiries || []

      // Auto-select first expiry
      if (expiries.value.length > 0 && !expiry.value) {
        expiry.value = expiries.value[0]
      }
    } catch (err) {
      console.error('[OFO] Failed to fetch expiries:', err)
      error.value = 'Failed to load expiries'
    }
  }

  /**
   * Toggle strategy selection
   */
  function toggleStrategy(strategyKey) {
    const idx = selectedStrategies.value.indexOf(strategyKey)
    if (idx === -1) {
      selectedStrategies.value.push(strategyKey)
    } else {
      selectedStrategies.value.splice(idx, 1)
    }
  }

  /**
   * Check if strategy is selected
   */
  function isStrategySelected(strategyKey) {
    return selectedStrategies.value.includes(strategyKey)
  }

  /**
   * Select all strategies
   */
  function selectAllStrategies() {
    selectedStrategies.value = availableStrategies.value.map(s => s.key)
  }

  /**
   * Clear all strategy selections
   */
  function clearStrategies() {
    selectedStrategies.value = []
  }

  /**
   * Calculate best strategy combinations
   */
  async function calculate() {
    if (!underlying.value || !expiry.value) {
      error.value = 'Please select underlying and expiry'
      return
    }

    if (selectedStrategies.value.length === 0) {
      error.value = 'Please select at least one strategy'
      return
    }

    isLoading.value = true
    error.value = null

    try {
      const response = await api.post('/api/ofo/calculate', {
        underlying: underlying.value,
        expiry: expiry.value,
        strategy_types: selectedStrategies.value,
        strike_range: strikeRange.value,
        lots: lots.value
      })

      const data = response.data
      results.value = data.results || {}
      spotPrice.value = data.spot_price || 0
      atmStrike.value = data.atm_strike || 0
      lotSize.value = data.lot_size || getLotSize(underlying.value)
      calculationTimeMs.value = data.calculation_time_ms || 0
      totalCombinationsEvaluated.value = data.total_combinations_evaluated || 0
      lastCalculated.value = data.calculated_at || new Date().toISOString()

      console.log(`[OFO] Calculated ${totalCombinationsEvaluated.value} combinations in ${calculationTimeMs.value}ms`)
    } catch (err) {
      console.error('[OFO] Calculation failed:', err)
      error.value = err.response?.data?.detail || 'Failed to calculate strategies'
      results.value = {}
    } finally {
      isLoading.value = false
    }
  }

  /**
   * Start auto-refresh timer
   */
  function startAutoRefresh() {
    stopAutoRefresh()
    autoRefreshEnabled.value = true
    autoRefreshTimer = setInterval(() => {
      if (selectedStrategies.value.length > 0 && expiry.value) {
        calculate()
      }
    }, autoRefreshInterval.value * 60 * 1000)
    console.log(`[OFO] Auto-refresh started: every ${autoRefreshInterval.value} minutes`)
  }

  /**
   * Stop auto-refresh timer
   */
  function stopAutoRefresh() {
    autoRefreshEnabled.value = false
    if (autoRefreshTimer) {
      clearInterval(autoRefreshTimer)
      autoRefreshTimer = null
    }
    console.log('[OFO] Auto-refresh stopped')
  }

  /**
   * Toggle auto-refresh
   */
  function toggleAutoRefresh() {
    if (autoRefreshEnabled.value) {
      stopAutoRefresh()
    } else {
      startAutoRefresh()
    }
  }

  /**
   * Set auto-refresh interval and restart if enabled
   */
  function setAutoRefreshInterval(minutes) {
    autoRefreshInterval.value = minutes
    if (autoRefreshEnabled.value) {
      startAutoRefresh()
    }
  }

  /**
   * Open strategy in Strategy Builder
   */
  async function openInStrategyBuilder(result) {
    const strategyStore = useStrategyStore()

    // Set underlying and expiry
    await strategyStore.setUnderlying(underlying.value)

    // Clear existing legs
    strategyStore.clearLegs()

    // Add each leg from the result
    for (const leg of result.legs) {
      strategyStore.addLegFromOFO({
        expiry_date: leg.expiry,
        contract_type: leg.contract_type,
        transaction_type: leg.transaction_type,
        strike_price: leg.strike,
        lots: leg.lots,
        entry_price: leg.cmp,
        instrument_token: leg.instrument_token,
        tradingsymbol: leg.tradingsymbol
      })
    }

    // Return route path for navigation
    return '/strategy'
  }

  /**
   * Get strategy display name
   */
  function getStrategyName(strategyKey) {
    const strategy = availableStrategies.value.find(s => s.key === strategyKey)
    return strategy?.name || strategyKey
  }

  /**
   * Reset store state
   */
  function reset() {
    stopAutoRefresh()
    underlying.value = 'NIFTY'
    expiry.value = ''
    expiries.value = []
    selectedStrategies.value = []
    strikeRange.value = 10
    lots.value = 1
    results.value = {}
    spotPrice.value = 0
    atmStrike.value = 0
    lotSize.value = getLotSize('NIFTY')
    calculationTimeMs.value = 0
    totalCombinationsEvaluated.value = 0
    lastCalculated.value = null
    isLoading.value = false
    error.value = null
  }

  // =========================================================================
  // RETURN
  // =========================================================================

  return {
    // State
    underlying,
    expiry,
    expiries,
    selectedStrategies,
    strikeRange,
    lots,
    results,
    spotPrice,
    atmStrike,
    lotSize,
    calculationTimeMs,
    totalCombinationsEvaluated,
    lastCalculated,
    autoRefreshEnabled,
    autoRefreshInterval,
    isLoading,
    error,
    availableStrategies,

    // Getters
    selectedCount,
    hasResults,
    orderedResults,
    lastCalculatedFormatted,

    // Actions
    setUnderlying,
    fetchExpiries,
    toggleStrategy,
    isStrategySelected,
    selectAllStrategies,
    clearStrategies,
    calculate,
    startAutoRefresh,
    stopAutoRefresh,
    toggleAutoRefresh,
    setAutoRefreshInterval,
    openInStrategyBuilder,
    getStrategyName,
    reset
  }
})

File: backend/app/schemas/ofo.py
"""
OFO (Options For Options) Schemas

Request/response schemas for the OFO feature that finds and ranks
the best strategy combinations from option chain data.
"""

from datetime import datetime
from typing import Dict, List, Optional
from pydantic import BaseModel, Field


class OFOCalculateRequest(BaseModel):
    """Request schema for OFO calculation."""

    underlying: str = Field(..., description="Underlying index: NIFTY, BANKNIFTY, FINNIFTY")
    expiry: str = Field(..., description="Expiry date in YYYY-MM-DD format")
    strategy_types: List[str] = Field(
        ...,
        description="List of strategy types to calculate: iron_condor, short_straddle, etc."
    )
    strike_range: int = Field(
        default=10,
        ge=5,
        le=20,
        description="Number of strikes to consider on each side of ATM (±5, ±10, ±15, ±20)"
    )
    lots: int = Field(
        default=1,
        ge=1,
        le=50,
        description="Number of lots per leg"
    )

    class Config:
        json_schema_extra = {
            "example": {
                "underlying": "NIFTY",
                "expiry": "2025-01-09",
                "strategy_types": ["iron_condor", "short_strangle", "bull_call_spread"],
                "strike_range": 10,
                "lots": 1
            }
        }


class OFOLegResult(BaseModel):
    """Individual leg details in OFO result."""

    expiry: str = Field(..., description="Expiry date")
    contract_type: str = Field(..., description="CE or PE")
    transaction_type: str = Field(..., description="BUY or SELL")
    strike: float = Field(..., description="Strike price")
    cmp: float = Field(..., description="Current market price (LTP)")
    lots: int = Field(..., description="Number of lots")
    qty: int = Field(..., description="Total quantity (lots × lot_size)")
    instrument_token: int = Field(..., description="Kite instrument token for order placement")
    tradingsymbol: str = Field(..., description="Trading symbol for order placement")

    class Config:
        json_schema_extra = {
            "example": {
                "expiry": "2025-01-09",
                "contract_type": "CE",
                "transaction_type": "SELL",
                "strike": 24200.0,
                "cmp": 150.50,
                "lots": 1,
                "qty": 25,
                "instrument_token": 12345678,
                "tradingsymbol": "NIFTY2510924200CE"
            }
        }


class OFOStrategyResult(BaseModel):
    """Single strategy combination result."""

    strategy_type: str = Field(..., description="Strategy type key (e.g., iron_condor)")
    strategy_name: str = Field(..., description="Display name (e.g., Iron Condor)")
    max_profit: float = Field(..., description="Maximum profit at expiry")
    max_loss: float = Field(..., description="Maximum loss at expiry")
    breakevens: List[float] = Field(default_factory=list, description="Breakeven points")
    net_premium: float = Field(..., description="Net premium received/paid (positive = credit)")
    risk_reward_ratio: float = Field(..., description="Max profit / Max loss ratio")
    legs: List[OFOLegResult] = Field(..., description="List of option legs")

    class Config:
        json_schema_extra = {
            "example": {
                "strategy_type": "iron_condor",
                "strategy_name": "Iron Condor",
                "max_profit": 4500.0,
                "max_loss": -8000.0,
                "breakevens": [23850.0, 24350.0],
                "net_premium": 180.0,
                "risk_reward_ratio": 0.56,
                "legs": []
            }
        }


class OFOCalculateResponse(BaseModel):
    """Response schema for OFO calculation."""

    underlying: str = Field(..., description="Underlying index")
    expiry: str = Field(..., description="Expiry date")
    spot_price: float = Field(..., description="Current spot price of underlying")
    atm_strike: float = Field(..., description="ATM strike price")
    lot_size: int = Field(..., description="Lot size for this underlying")
    calculated_at: datetime = Field(..., description="Timestamp of calculation")
    calculation_time_ms: int = Field(..., description="Time taken to calculate in milliseconds")
    total_combinations_evaluated: int = Field(
        default=0,
        description="Total number of combinations evaluated across all strategies"
    )
    results: Dict[str, List[OFOStrategyResult]] = Field(
        ...,
        description="Results grouped by strategy type, each containing top 3 combinations"
    )
    data_freshness: str = Field(
        default="LIVE",
        description="'LIVE' when market is open, 'LAST_KNOWN' when showing previous session's close prices"
    )

    class Config:
        json_schema_extra = {
            "example": {
                "underlying": "NIFTY",
                "expiry": "2025-01-09",
                "spot_price": 24100.50,
                "atm_strike": 24100.0,
                "lot_size": 25,
                "calculated_at": "2025-01-02T10:30:00",
                "calculation_time_ms": 1250,
                "total_combinations_evaluated": 15420,
                "results": {
                    "iron_condor": [],
                    "short_strangle": [],
                    "bull_call_spread": []
                }
            }
        }


# Available strategies for OFO (9 popular strategies)
OFO_AVAILABLE_STRATEGIES = [
    {
        "key": "iron_condor",
        "name": "Iron Condor",
        "category": "neutral",
        "legs_count": 4,
        "description": "Bull put spread + Bear call spread for range-bound trading"
    },
    {
        "key": "iron_butterfly",
        "name": "Iron Butterfly",
        "category": "neutral",
        "legs_count": 4,
        "description": "ATM short straddle with protective wings"
    },
    {
        "key": "short_straddle",
        "name": "Short Straddle",
        "category": "neutral",
        "legs_count": 2,
        "description": "Sell ATM CE + PE at same strike"
    },
    {
        "key": "short_strangle",
        "name": "Short Strangle",
        "category": "neutral",
        "legs_count": 2,
        "description": "Sell OTM CE + PE at different strikes"
    },
    {
        "key": "long_straddle",
        "name": "Long Straddle",
        "category": "volatile",
        "legs_count": 2,
        "description": "Buy ATM CE + PE at same strike"
    },
    {
        "key": "long_strangle",
        "name": "Long Strangle",
        "category": "volatile",
        "legs_count": 2,
        "description": "Buy OTM CE + PE at different strikes"
    },
    {
        "key": "bull_call_spread",
        "name": "Bull Call Spread",
        "category": "bullish",
        "legs_count": 2,
        "description": "Buy lower CE, sell higher CE"
    },
    {
        "key": "bear_put_spread",
        "name": "Bear Put Spread",
        "category": "bearish",
        "legs_count": 2,
        "description": "Buy higher PE, sell lower PE"
    },
    {
        "key": "butterfly_spread",
        "name": "Butterfly Spread",
        "category": "neutral",
        "legs_count": 4,
        "description": "Buy low + high strikes, sell 2x middle strike"
    }
]
